onstart.go

```diff
@@ -0,0 +1,188 @@
+//    }),
+//    OnStop(func(ctx context.Context, server *http.Server) error {
+//      return server.Shutdown(ctx)
+//    }),
+//  )
+func OnStart(funcs ...interface{}) Option {
+	var opts []Option
+	for _, f := range funcs {
+		opts = append(opts, onStart(reflect.ValueOf(f)))
+	}
+	return Options(opts...)
+}
+
+func validateOnStartFunc(ft reflect.Type) (onStartType, error) {
+	if kind := ft.Kind(); kind != reflect.Func {
+		return onStartType{}, fmt.Errorf("OnStart must be passed a function, got %v", kind)
+	}
+
+	onStartDesc := onStartType{
+		hasCtx: ft.NumIn() > 0 && ft.In(0) == _ctxType,
+	}
+
+	switch ft.NumOut() {
+	case 0:
+		// No return value, nothing to validate.
+	case 1:
+		switch ft.Out(0) {
+		case _errType:
+			onStartDesc.hasErr = true
+		case _stopperType:
+			onStartDesc.hasStopper = true
+		default:
+			return onStartDesc, fmt.Errorf("OnStart functions can only return an error or a Stopper, found %v", ft.Out(0))
+		}
+	case 2:
+		if ft.Out(0) != _stopperType || ft.Out(1) != _errType {
+			return onStartDesc, fmt.Errorf("OnStart functions can only return (Stopper, error), found (%v, %v)", ft.Out(0), ft.Out(1))
+		}
+		onStartDesc.hasErr = true
+		onStartDesc.hasStopper = true
+	default:
+		return onStartDesc, fmt.Errorf("OnStart functions must not have more than 2 returns")
+	}
+
+	return onStartDesc, nil
+}
+
+func onStart(f reflect.Value) Option {
+	ft := f.Type()
+
+	onStartDesc, err := validateOnStartFunc(ft)
+	if err != nil {
+		return invokeErr(err)
+	}
+
+	// The invoke function will take a lifecycle, followed by any arguments to f.
+	invokeInTypes := []reflect.Type{_lifecycleType}
+	for i := 0; i < ft.NumIn(); i++ {
+		if i == 0 && onStartDesc.hasCtx {
+			continue
+		}
+
+		invokeInTypes = append(invokeInTypes, ft.In(i))
+	}
+
+	invokeFuncType := reflect.FuncOf(invokeInTypes, []reflect.Type{_errType}, false /* variadic */)
+
+	invokeFunc := reflect.MakeFunc(invokeFuncType, func(args []reflect.Value) []reflect.Value {
+		// The first argument to invoke is a lifecycle, let's extract that.
+		lifecycle := args[0].Interface().(Lifecycle)
+		var onStop Stopper
+		lifecycle.Append(Hook{
+			OnStart: func(ctx context.Context) error {
+				var startArgs []reflect.Value
+
+				if onStartDesc.hasCtx {
+					startArgs = append(startArgs, reflect.ValueOf(ctx))
+				}
+
+				startArgs = append(startArgs, args[1:]...)
+				res := f.Call(startArgs)
+
+				// If the OnStop returns an error, then it must be the last argument
+				if onStartDesc.hasErr {
+					if err := res[len(res)-1].Interface(); err != nil {
+						return err.(error)
+					}
+				}
+
+				// If the OnStop returns a Stopper, it must be the first argument.
+				if onStartDesc.hasStopper {
+					onStop = res[0].Interface().(Stopper)
+				}
+
+				return nil
+			},
+			OnStop: func(ctx context.Context) error {
+				if onStop == nil {
+					return nil
+				}
+
+				return onStop(ctx)
+			},
+		})
+
+		return []reflect.Value{reflect.New(_errType).Elem()}
```

Comment:
Can use reflect.Zero
```suggestion
		return []reflect.Value{reflect.Zero(_errType)}
```